File: clocks.py
"""
ID: wrwwctb1
LANG: PYTHON3
TASK: clocks
6.5.4

review

constant time sol exists

num of all distinct combinations of moves = 4**9 (order doesn't matter)


my sol doesn't use the above

num of all distinct states = 4**9

states, moves and state transitions are coded in bits
"""

import os, sys, re, time
from collections import Counter, deque, defaultdict
from queue import Queue
from copy import copy, deepcopy
from itertools import combinations, permutations, accumulate, \
                      combinations_with_replacement
from functools import lru_cache, cmp_to_key, partial as functools_partial
from heapq import heappush, heappop, nlargest, nsmallest
from bisect import bisect_left, bisect_right
from math import ceil, floor, factorial, gcd, modf, log, log2, log10, sqrt, \
         pi, sin, cos, tan, asin, acos, atan, atan2, hypot, erf, erfc, inf, nan
# sys.setrecursionlimit(5782)
# python -m cProfile -s time ha.py

print = functools_partial(print, flush=True)
input = lambda: fin.readline().strip('\n')

def printwrite(string):
    print(string)
    fout.write(str(string) + '\n')

filename = 'clocks'
fin = open(filename + '.in', 'r')
fout = open(filename + '.out', 'w')









def clocks2bits(clocks):
    num = 0
    for i in range(len(clocks)):
        num += (clocks[i] // 3 % 4) * (1 << (i << 1))
    return num

def moveLet2bits(moveLet):
    move = 0
    for ch in moveLet:
        d = ord(ch) - ord('A')
        move += 1 << 2*d
    return move

def transition(clocks, move):
    b0 = clocks & 0b010101010101010101
    b1 = clocks & 0b101010101010101010
    a0 = b0 ^ move
    a1 = b1 ^ ((b0 & move) << 1)
    return a0 + a1


clocks = []
for i in range(3):
    row = list(map(int, input().split()))
    clocks.extend(row)
clocks = clocks2bits(clocks)

moveLets = ['ABDE',
            'ABC',
            'BCEF',
            'ADG',
            'BDEFH',
            'CFI',
            'DEGH',
            'GHI',
            'EFHI']

moves = []
for moveLet in moveLets:
    moves.append(moveLet2bits(moveLet))

#tests = [3, 4, 7, 8]
##tests = [4, 7, 3, 8]
#print(bin(clocks))
#for test in tests:
#    clocks = transition(clocks, moves[test])
#    print(bin(clocks))

seen = [False] * (4**9)
seen[clocks] = True
qq = deque()
qq.append(clocks)

parent = [-1] * (4**9)
arrivedVia = [-1] * (4**9)

# bfs
while qq:
    u = qq.popleft()
    if u == 0: # destination
        break
    for i, move in enumerate(moves):
        v = transition(u, move)
        if seen[v]:
            continue
        seen[v] = True
        qq.append(v)
        parent[v] = u
        arrivedVia[v] = i

# trace
v = 0
used = []
while True:
    u = parent[v]
    if u == -1:
        break
    used.append(arrivedVia[v] + 1)
    v = u

printwrite(' '.join(['%d'] * len(used)) % tuple(used[::-1]))












fin.close()
fout.close()
del print, input

File: clocks.cpp
/*
ID: wrwwctb1
TASK: clocks
LANG: C++14
*/

#include <iostream>
#include <string>
#include <fstream>
#include <sstream>
#include <cassert>

#include <vector>
#include <deque>
#include <list>
#include <set> // has multiset
#include <map> // has multimap
#include <unordered_set> // has unordered_multiset
#include <unordered_map> // has unordered_multimap
#include <stack>
#include <queue> // has priority_queue

#include <algorithm>
#include <cmath>
#include <cfloat>
#include <ctime>
#include <climits>
#include <iomanip>
#include <cstring>

#define FOR( i, N ) for( int (i) = 0; (i) < (N); (i)++ )
//#define FOR_( i, start, endplusone ) for( int (i) = (start); (i) < (endplusone); (i)++ )
const double EPS = 1e-6;
const double PI = 3.141592653589793;
using namespace std;
// profile: 27196474















int clocks2bits(int clocks[],int C){
    int num = 0;
    FOR (i, C)
        num += (clocks[i] / 3 % 4) * (1 << (i << 1));
    return num;
}

int moveLet2bits(string moveLet){
    int move = 0;
    for (char ch: moveLet)
        move += 1 << 2*(ch - 'A');
    return move;
}

int transition(int clocks, int move){
    int b0 = clocks & 0b010101010101010101;
    int b1 = clocks & 0b101010101010101010;
    int a0 = b0 ^ move;
    int a1 = b1 ^ ((b0 & move) << 1);
    return a0 + a1;
}

int main(int argc, char **argv){
    time_t ttt = time(nullptr);


    const string filename("clocks");




    string filenameo = filename + ".out";
    string filenamei = filename + ".in";
    ofstream fout(filenameo.c_str());
    ifstream fin(filenamei.c_str());






    const int C = 9;
    int clocksDig[C];
    FOR (c, C)
        fin >> clocksDig[c];
    int clocks = clocks2bits(clocksDig, C);

    const int M = 9;
    string moveLets[M] = {"ABDE",
                          "ABC",
                          "BCEF",
                          "ADG",
                          "BDEFH",
                          "CFI",
                          "DEGH",
                          "GHI",
                          "EFHI"};

    int moves[M];
    FOR (m, M)
        moves[m] = moveLet2bits(moveLets[m]);

    const int Nstates = 1 << 18;

    bool seen[Nstates] = {};
    seen[clocks] = true;
    queue<int> qq;
    qq.push(clocks);

    int parent[Nstates];
    int *arrivedVia = new int[Nstates]; // SIGSEGV if using array. stack overflow

    memset(parent, -1, sizeof(parent));

    while (!qq.empty()){
        int u = qq.front();
        qq.pop();
        if (u == 0)
            break;
        FOR (i, M){
            int v = transition(u, moves[i]);
            if (seen[v])
                continue;
            seen[v] = true;
            qq.push(v);
            parent[v] = u;
            arrivedVia[v] = i;
        }
    }

    int v = 0;
    vector<int> used;
    while (true){
        int u = parent[v];
        if (u == -1)
            break;
        used.push_back(arrivedVia[v] + 1);
        v = u;
    }

    FOR (i, used.size()){
        fout << used[used.size() - 1 - i];
        if (i != used.size()-1)
            fout << ' ';
    }
    fout << endl;








    fin.close();
    fout.close();


    printf("~%d\n", time(nullptr) - ttt);
    // system("pause");
    return 0;
}